src/OrchardCore/OrchardCore.DisplayManagement.Liquid/LiquidViewTemplate.cs

```diff
@@ -184,21 +192,33 @@ public static void Contextualize(this TemplateContext context, RazorPage page, o
             var layout = layoutAccessor.GetLayout();
             context.AmbientValues.Add("ThemeLayout", layout);
 
-            var site = await services.GetRequiredService<ISiteService>().GetSiteSettingsAsync();
-            context.MemberAccessStrategy.Register(site.GetType());
-            context.LocalScope.SetValue("Site", site);
+            // TODO: Extract the request culture
+
+            var values = new Dictionary<string, object>();
+            var providers = services.GetServices<ILiquidValueProvider>();
 
-            context.LocalScope.SetValue("User", displayContext.ViewContext.HttpContext.User);
+            foreach (var provider in providers)
+            {
+                await provider.PopulateValuesAsync(values);
+            }
 
-            // TODO: Extract the request culture instead of the default site's one
-            if (site.Culture != null)
+            foreach (var value in values)
             {
-                context.CultureInfo = new CultureInfo(site.Culture);
+                var type = value.Value.GetType();
+
+                if (Type.GetTypeCode(type) == TypeCode.Object)
```

Comment:
Meh, why not expose the context directly in the `ILiquidValueProvider` instead? Then we could just extend anything, and add the type if necessary. Then the module can decide if the type should be registered globally or on the context.

Reply:
Yes, simpler, i was not sure we can allow this.

> Then the module can decide if the type should be registered globally or on the context

You also say globally, do you mean e.g that `O.Users` could register `ClaimsPrincipal` globally? Here or in a static constructor? Anyway, when registering globally, as i remember, we would need to consider concurrent writing.